packages/di/src/lib/injector.ts

```diff
@@ -63,8 +73,8 @@ export class Injector {
     this.#instances = new WeakMap();
   }
 
-  #createAndCache<T extends Injectable>(token: ProviderToken<T>): T {
-    const instance = new token();
+  #createAndCache<T extends Injectable>(token: ProviderToken<T>, factory: () => T): T {
+    const instance = factory();
```

Comment:
I would also call the factory passing the injector `factory(this)`. Will it recurse when accessing the same Token? I guess it should provide the parent service instead in that particular case.
This facilitates easy decoration on demand inside the factory.